### PR TITLE
Refactored adapters and use cases layers

### PR DESCRIPTION
**Pull request by enerko**
Main changes I made:
- Added input and output boundaries
- Moved the game thread over to the use case interactor, UpdatePlayer
- Made all static methods in Player and UpdatePlayer non-static
- Removed GamePanel, and moved the functionality to main

By implementing the boundaries, I ensured that the presenter and the controller both depend on abstraction, instead of the use case interactor itself. To see a simplified process of the controller presenter interaction, check this document out: [https://docs.google.com/document/d/1O9iOUfxDMZELKMQhlcU0RpRVQfxMt0f7QJ9r09bdVQA/edit?usp=sharing](url).

Right now the game feels laggy, and I think the issue is because the game continues to receive input even when the player positions are not being updated. To fix this, we could ensure that the game only takes in input when it should be updating the player position. 

If you have any suggestions or ideas, feel free to comment :)

### CODE REVIEW

File: src/main/java/use_cases/default_game/UpdatePlayer.java [hunk 1]
```diff
@@ -2,53 +2,99 @@
...
-    public static int playerSpeed = 48; // might move this to controller
+public class UpdatePlayer extends JPanel implements IGamePanelInputBoundary, Runnable {
+
+    public Thread gameThread;
+    private final Player player;
+    private final IGamePanelOutputBoundary outputBoundary;
+    final int SPRITE_TILE_SIZE = 16;
+    final int SCALE = 3; // may be changed to an unfixed variable later
+    final int TILE_SIZE = SPRITE_TILE_SIZE * SCALE;
+    private final int FPS = 20; // updates the screen 5 times per second
```
[the remark this answers is not in the document]
Comment (enerko) [reply src/main/java/use_cases/default_game/UpdatePlayer.java]: I made it 20 for now to make the game feel less laggy (player looks like it's teleporting with FPS = 5). The only thing that the FPS changes is how many times it updates the player position and doesn't actually change the player speed. We need to actually stop the game from receiving user input within certain intervals to change the player speed.

File: src/main/java/use_cases/default_game/UpdatePlayer.java [hunk 2]
```diff
@@ -2,53 +2,99 @@
...
+     * Moves the player to the direction specified by the user
+     * input keycode.
+     *
+     * @param keycode the keyboard input
      */
-    public static void movePlayerDown() {
-        Player.movePlayerY(playerSpeed);
+    public void movePlayer(int keycode) {
+        if (keycode == KeyEvent.VK_W) {
+            player.movePlayerY(-TILE_SIZE);
```
[the remark this answers is not in the document]
Comment (enerko) [reply src/main/java/use_cases/default_game/UpdatePlayer.java]: Should be good now. I will keep the class name UpdatePlayer the same for now as I don't know if we are updating anything else other than player position in that class.

**Comment by ltenenb** [comment]
Right now the body of the `while (gameThread != null) { ... }` loop gets run over and over again as fast as possible, even though it only needs to call update() every 50ms. This means that our program's CPU usage will be much higher than it needs to be. I would suggest instead doing:
```java
long lastTime = System.currentTimeMillis();
while (gameThread != null) {
    long currentTime = System.currentTimeMillis();
    long sleepTime = lastTime + 1000 / FPS - currentTime;
    if (sleepTime > 0)
        Thread.sleep(sleepTime);
    lastTime = currentTime;
    update();
}
```

Also, I think the default implementations in `IGameInput/OutputBoundary` can be removed. I'm not sure why they're there.

Other than that, this looks good!

**Comment by enerko** [comment]
Implemented the requested changes. The game is still laggy and seems like KeyEventListener is receiving user input non-stop. Since it is not exactly an error that needs to be fixed immediately, I will resolve this issue later in another pull request.

**Comment by tseamuscorlett** [comment]
> Implemented the requested changes. The game is still laggy and seems like KeyEventListener is receiving user input non-stop. Since it is not exactly an error that needs to be fixed immediately, I will resolve this issue later in another pull request.

Do you think implementing KeyReleased would fix it?

**Review by tseamuscorlett — APPROVED** [review]

**Comment by enerko** [comment]
> > Implemented the requested changes. The game is still laggy and seems like KeyEventListener is receiving user input non-stop. Since it is not exactly an error that needs to be fixed immediately, I will resolve this issue later in another pull request.
> 
> Do you think implementing KeyReleased would fix it?

 Since the problem happens when the user is pressing the key, I don't think that would solve the issue. One option would be to move the player one tile at a time, instead of moving it while the user is pressing the key. But this implementation would be annoying for the user I assume.

File: src/main/java/adapters/default_game/GamePanelPresenter.java [hunk 3]
```diff
@@ -104,6 +69,9 @@ public Graphics2D graphics() {
                 return g2;
             }
         };
+        MazeItems items = new MazeItems();
+        MazeHazards hazards = new MazeHazards();
+        new CustomAssetSetter("mazes/maze02.txt", items, hazards);
```
Comment (tseamuscorlett) [inline src/main/java/adapters/default_game/GamePanelPresenter.java:74]: I think this is the problem. Every time paintComponent is called, it's constructing a new MazeItems, MazeHazards, AssetSetter and drawing the map. With this set up we can't reflect changes to the Hazards/Items. The game keeps drawing the original setup.
This is why I think when the Controller & Presenter are initiated, they should take in the same Items & Hazards

**Pull request merged**